feat(backend): access control

disKeith opened the pull request:
## Context

This PR adds simple access control to limit users' access from all templates on the product

Closes [[Notion Task](https://www.notion.so/opengov/2ccadbea5c734aa09614b4ca8cc7177e?v=76cff9fca57b496f857c990e2639af2f&p=d1b56d2e1f94483489eab59ad03029d7&pm=s)]

## Approach

Implement a new `UserTemplateAccess` entity that establishes an implicit many to many relationship between `Template` and `User`
- Each row contains a `userId` and `templateId` that denotes the specific user having access to the specific template
- Allows for future extensions e.g. `access_type` column for fine-grain access

**Features**:

1. `AccessControlService`
   - `createAccess`: adds new row in `access` table if the access doesn't exist
   - `addAccess`: [FF] checks if current user has access to the template, find/create the specified user to invite, and calls `createAccess` to create a new row if it doesn't exist
   - `findAccess`: [FF] checks if the parsed `userId` and `templateId` access exists (returns true), or return true if feature-flag is off
   - `listAccessByUserId`: [FF] returns the list of templates that the specified `userId` can access, or all templates if feature-flag is off

2. `AccessControlController`
   - Exposes `{{host}}/api/access-control/add` endpoint that can be used to add template access to a specified `userEmailToInvite`. Called with body:
```
{
    "templateId": <id>,
    "userEmailToInvite": <email-string>
}
``` 

3. `LettersController` & `NotificationsController`
   - added `userId` in parameter when calling `templatesService.findOne`

4. `TemplatesService`
   - `create`: added transaction to create access during template creation
   - `findAll`: updated to retrieve and display only the allowed templates to the user
   - `findOne`: updated to check if user has access to template

5. `TemplatesController`
   - added `userId` in parameter when calling `templatesService.findAll` and `templatesService.findOne`

disKeith commented on backend/src/access-control/access-control.service.ts:
```diff
@@ -0,0 +1,141 @@
+      return accessList
+    }
+
+    const accessList = await this.accessControlRepository.find({
+      where: {
+        userId: userId,
+      },
+      relations: {
+        user: true,
+        template: true,
+      },
+    })
```
tried using the `order` option within the `find` method didn't work for me as the column we want to sort by (`template.createdAt`) is not assignable to the entity's type:

```
const accessList = await this.accessControlRepository.find({
      where: {
        userId: userId,
      },
      relations: {
        template: true,
      },
      order: {
        'template.createdAt': 'DESC',
      },
    })
```

here's a working solution by explicity defining the query:

```
const accessList = await this.accessControlRepository
      .createQueryBuilder('access')
      .leftJoinAndSelect('access.template', 'template')
      .where('access.userId = :userId', { userId: userId })
      .orderBy('template.createdAt', 'DESC')
      .getMany()
```

wdyt?

disKeith commented on backend/src/templates/templates.service.ts:
```diff
@@ -39,19 +42,34 @@ export class TemplatesService {
+  async findAll(userId: number) {
+    const templates = await this.accessControlService.listAccessByUserId(userId)
+
+    return templates.sort((a, b) => {
+      return +new Date(b.createdAt) - +new Date(a.createdAt)
```
https://github.com/opengovsg/letters/pull/295#discussion_r1332433408

disKeith commented on backend/src/access-control/access-control.service.ts:
```diff
@@ -0,0 +1,141 @@
+
+    const userIdToInvite = userToInvite.id
+
+    if (await this.findAccess(userIdToInvite, templateId)) {
+      throw new BadRequestException(
+        `access for user ${userIdToInvite} and template ${templateId} already exists`,
+      )
+    }
```
hmm let me see if i understood you correctly, the `createAccess` function is directly called only when we create a new template in which case we shouldn't be faced with the scenario where we might add duplicated pairings, in contrast to `addAccess` which should check for existing records as it might be called with an existing pairing in the table

disKeith commented on backend/src/access-control/access-control.service.ts:
```diff
@@ -0,0 +1,141 @@
+
+    if (!userToInvite) {
+      throw new BadRequestException(
+        `user with email ${userEmailToInvite} not found`,
```
seems okay to me, since the `email` column is decorated with `@IsGovSgEmail`

disKeith commented on backend/src/access-control/access-control.service.ts:
```diff
@@ -0,0 +1,141 @@
+    const userToInvite = await this.usersRepository.findOne({
+      where: { email: userEmailToInvite },
+      select: ['id'],
+    })
```
oh yes you're right! thanks for pointing this out~ https://github.com/opengovsg/letters/pull/295/commits/a6c7d11231fcbb97ebf988dd639d091d7c01a760

disKeith commented on backend/src/access-control/access-control.service.ts:
```diff
@@ -0,0 +1,141 @@
+        templateId: templateId,
+      },
+      relations: {
+        user: true,
```
right, that's redundant, updated! https://github.com/opengovsg/letters/pull/295/commits/a6c7d11231fcbb97ebf988dd639d091d7c01a760

disKeith commented on backend/src/access-control/access-control.service.ts:
```diff
@@ -0,0 +1,141 @@
+      return true
+    }
+
+    const accessExists = await this.accessControlRepository.findOne({
```
ooh great find! updated with this change https://github.com/opengovsg/letters/pull/295/commits/a6c7d11231fcbb97ebf988dd639d091d7c01a760

disKeith commented on backend/src/access-control/access-control.service.ts:
```diff
@@ -0,0 +1,141 @@
+      return true
+    }
+
+    return false
```
neat! https://github.com/opengovsg/letters/pull/295/commits/a6c7d11231fcbb97ebf988dd639d091d7c01a760

gweiying commented on backend/src/templates/templates.service.ts:
```diff
@@ -39,19 +42,34 @@ export class TemplatesService {
     }
-    return await this.repository.save(templateWithSampleLetter)
+
+    return await this.dataSource.transaction(async () => {
+      await this.accessControlService.createAccess(userId, createdTemplate.id)
+      return await this.repository.save(templateWithSampleLetter)
+    })
```
Small issue here, I think you'll find that your transaction isn't working as expected. You must be using the `transactionalEntityManager` exposed by the `dataSource.transaction` wrapper to save the entities, e.g.
```
return await this.dataSource.transaction(async (transactionalEntityManager) => {
// code to save here e.g. this.accessControlService.createAccessWithTransaction(userId, createdTemplate.id, transactionalEntityManager)
...code to save template
}
```



Above we're still using the global entity manager, so I think the transaction will not work as expected. Can see [here](https://orkhan.gitbook.io/typeorm/docs/transactions#creating-and-using-transactions) for more details on usage

gweiying commented on backend/src/templates/templates.service.ts:
```diff
@@ -39,19 +42,34 @@ export class TemplatesService {
+    if (hasTemplateAccess)
+      return await this.repository.findOne({ where: { id }, ...options })
+
+    return null
```
why not directly throw an error here?

gweiying commented on backend/src/access-control/access-control.service.ts:
```diff
@@ -0,0 +1,127 @@
+    )
+
+    if (!isAccessControlEnabled) {
+      const accessList = await this.templatesRepository.find()
```
small nit, but access control service shouldn't be using the templates repository directly - would be probably preferable to call template service instead!

gweiying commented on backend/src/access-control/access-control.service.ts:
```diff
@@ -0,0 +1,127 @@
+  ) {}
+
+  async createAccess(userId: number, templateId: number) {
+    const userExists = await this.usersRepository.findOne({
+      where: { id: userId },
+    })
```
small nit, but access control service shouldn't be using the users repository directly - would be probably preferable to call auth service instead

gweiying commented on backend/src/access-control/access-control.service.ts:
```diff
@@ -0,0 +1,127 @@
+    const userExists = await this.usersRepository.findOne({
+      where: { id: userId },
+    })
+    const templateExists = await this.templatesRepository.findOne({
+      where: { id: templateId },
+    })
```
small nit, but access control service shouldn't be using the templates repository directly - would be probably preferable to call template service instead!

gweiying reviewed (changes requested):
left some minor nits, might need to look specifically at the transaction manager

disKeith commented:
> left some minor nits, might need to look specifically at the transaction manager

addressed fixes for the comments in https://github.com/opengovsg/letters/pull/295/commits/50656ffa22c1f969d5285f594227af3f59f49e5f

from our offline discussion, have updated the logic flow for access control to be called only from `TemplatesService` and moved the endpoint for the add access function to be called from `TemplatesController` instead (via `{{host}}/id/invite`) with body `{"userEmail": "email-string"}`

disKeith commented on backend/src/templates/templates.service.ts:
```diff
@@ -39,19 +62,44 @@ export class TemplatesService {
+    if (isAccessControlEnabled) {
+      const templateIds = await this.accessControlService.listAccessByUserId(
+        userId,
+      )
```
ah yes indeed thanks! moved in https://github.com/opengovsg/letters/pull/295/commits/455eb69fe693451e3c11fb2158fbf595ce804a37

gweiying commented on backend/src/templates/templates.service.ts:
```diff
@@ -39,19 +62,38 @@ export class TemplatesService {
+          where: { id: In(templateIds) },
+          order: { createdAt: 'DESC' },
+        })
+      : await this.repository.find({
```
minor nit, maybe add a comment here why this would be null!

gweiying commented on backend/src/access-control/access-control.service.ts:
```diff
@@ -0,0 +1,129 @@
+    return entityManager.save([access])
+  }
+
+  async createAccess(
+    createAccessDto: CreateAccessDto,
+  ): Promise<UserTemplateAccess[]> {
+    const { userId, templateId } = createAccessDto
+    const userExists = await this.authService.exist({
+      where: { id: userId },
+    })
+
+    if (!userExists) {
+      throw new BadRequestException(`User with id ${userId} not found`)
+    }
+
+    const newAccess = new UserTemplateAccess()
+    newAccess.userId = userId
+    newAccess.templateId = templateId
+    return await this.accessControlRepository.save([newAccess])
+  }
```
can we subsume this function into add access so it's not accidentally used for now? so there'll only be createWithTransaction (used on template creation) and addAccess (used for template invitation)

disKeith replied:
sounds good! updated in https://github.com/opengovsg/letters/pull/295/commits/472cd10c018dfe6bca50bb3d82ec297c868924dc

gweiying reviewed (approved):
lgtm! works great, left inline suggestions on documentation or minor cleanups but performs beautifully!
